drivers/led/lp50xx.c

```diff
@@ -0,0 +1,371 @@
+	memcpy(data->chan_buf + 1, buf, num_channels);
+
+	return i2c_write_dt(&config->bus, data->chan_buf, num_channels + 1);
+}
+
+static int lp50xx_init(const struct device *dev)
+{
+	const struct lp50xx_config *config = dev->config;
+	uint8_t buf[3];
+	int err;
+
+	if (!device_is_ready(config->bus.bus)) {
+		LOG_ERR("%s: I2C device not ready", dev->name);
+		return -ENODEV;
+	}
+
+	if (config->num_leds > config->max_leds) {
+		LOG_ERR("%s: invalid number of LEDs %d (max %d)",
+			dev->name,
+			config->num_leds,
+			config->max_leds);
+		return -EINVAL;
+	}
+
+	/*
+	 * Since the status of the LP503x controller is unknown when entering
+	 * this function, and since there is no way to reset it, then the whole
+	 * configuration must be applied.
+	 */
+
+	/* Disable bank control for all LEDs. */
+	buf[0] = LP50XX_LED_CONFIG0;
+	buf[1] = 0;
+	buf[2] = 0;
+	if (config->chip == LP5009 || config->chip == LP5012) {
+		err = i2c_write_dt(&config->bus, buf, 2);
+	} else {
+		err = i2c_write_dt(&config->bus, buf, 3);
+	}
+	if (err < 0) {
+		return err;
+	}
+
+	/* Enable LED controller. */
+	buf[0] = LP50XX_DEVICE_CONFIG0;
+	buf[1] = CONFIG0_CHIP_EN;
+	err = i2c_write_dt(&config->bus, buf, 2);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Apply configuration. */
+	buf[0] = LP50XX_DEVICE_CONFIG1;
+	buf[1] = CONFIG1_PWM_DITHERING_EN | CONFIG1_AUTO_INCR_EN
+		| CONFIG1_POWER_SAVE_EN;
+	if (config->max_curr_opt) {
+		buf[1] |= CONFIG1_MAX_CURRENT_OPT;
+	}
+	if (config->log_scale_en) {
+		buf[1] |= CONFIG1_LOG_SCALE_EN;
+	}
+
+	return i2c_write_dt(&config->bus, buf, 2);
+}
+
+static const struct led_driver_api lp50xx_led_api = {
+	.on		= lp50xx_on,
+	.off		= lp50xx_off,
+	.get_info	= lp50xx_get_info,
+	.set_brightness	= lp50xx_set_brightness,
+	.set_color	= lp50xx_set_color,
+	.write_channels	= lp50xx_write_channels,
+};
+
+#define COLOR_MAPPING(led_node_id)						\
+	const uint8_t color_mapping_##led_node_id[] =				\
+		DT_PROP(led_node_id, color_mapping);
+
+#define LED_INFO(led_node_id)							\
+	{									\
+		.label		= DT_PROP(led_node_id, label),			\
+		.index		= DT_PROP(led_node_id, index),			\
+		.num_colors	=						\
+			DT_PROP_LEN(led_node_id, color_mapping),		\
+		.color_mapping	= color_mapping_##led_node_id,			\
+	},
+
+#define LP50XX_DEVICE(n, id, nchans, bbase_reg, cbase_reg)			\
+	DT_INST_FOREACH_CHILD(n, COLOR_MAPPING)					\
+										\
+	static const struct led_info lp##id##_leds_##n[] = {			\
+		DT_INST_FOREACH_CHILD(n, LED_INFO)				\
+	};									\
+										\
+	static const struct lp50xx_config lp##id##_config_##n = {		\
+		.bus			= I2C_DT_SPEC_INST_GET(n),		\
+		.chip			= LP##id,				\
+		.max_leds		= LP##id##_MAX_LEDS,			\
+		.num_leds		= ARRAY_SIZE(lp##id##_leds_##n),	\
+		.num_channels		= nchans,				\
+		.log_scale_en		= DT_INST_PROP(n, log_scale_en),	\
+		.max_curr_opt		= DT_INST_PROP(n, max_curr_opt),	\
+		.brightness_base_reg	= bbase_reg,				\
+		.color_base_reg		= cbase_reg,				\
+		.leds_info		= lp##id##_leds_##n,			\
```

Comment:
Wow, it is big ! `max_leds`, `num_channels`, `brightness_base_reg` and `color_base_reg` are not configuration information. These values ​​are shared/duplicated between all LEDs. Please let's optimize the memory usage a bit and get them out of the lp50xx_config structure.

Reply:
OK, I will remove it for next push using functions with the chip type instead.

Reply:
To remove `num_channels`, I need to populate the `chan_buf` field in `lp50xx_init()`. Is it OK to use k_malloc() inside the driver for this ?

Reply:
No you don't need to. Since the number of channels can be deduced from the chip type, then you can use a macro to statically initialize the channel buffer. I think it was done this way in #46134.

Reply:
#46134 did actually use a macro but with many ternary operators (and not all devices were supported). I could use a function like I did for the other informations but it won't work statically for initializing the buffer.
Would it be bad practice to initialize the buffer with the `LP5030_NUM_CHANNELS` value to make it work for all devices ?

Reply:
If I am correct the channel buffer size would be 49 bytes to work for all devices, while `LP5012` only needs a 13 bytes buffer. So yes it is not good. We have to put memory usage and efficiency first. I don't like complicated macros either, but hopefully you can make them understandable and beautiful.

Reply:
OK, I chose the easiest way out by defining all `LP50xx_NUM_CHANNELS` values. This will work for now while witing for a better solution to the `write_channels()` API.

Reply:
Good for me.